docs/it/refactoring.md

```diff
@@ -0,0 +1,161 @@
+In questi casi, ma più in generale, è necessario tenere a mente che l'attività di refactoring ha tra gli obiettivi quello di rivedere il codice rendendolo più mantenibile, nonché rifattorizzabile.
+Diretta conseguenza di questo obiettivo, è la sua applicazione nella pratica del TDD (Test Driven Development), dove il refactoring copre uno degli step.
+
+## Come
+
+Come si può intuire, con le attività di refactoring andremo a migliorare il codice in modo da renderlo meno complesso, più mantenibile, più resiliente.
+Per farlo, potremmo iniziare con:
+
+* Red-Green-Refactor;
+* Rinominare i metodi oscuri;
+* Cambiare la firma dei metodi;
+* Estrarre nuovi metodi;
+* Semplificare i metodi esistenti;
+* Astrarre.
+
+### Red-Green-Refactor
+
+Forse il metodo più popolare, utilizzato per mettere in pratica il TDD (Test Driven Development), dove chi sviluppa suddivide il processo di scrittura del codice in tre fasi:
+
+* Analizza quali sono le esigenze e scrive dapprima il test, che, per ovvi motivi, fallirà (_Red_);
+* Scrive il codice necessario affinché il test passi (_Green_);
+* Lo migliora (_Refactoring_)
+
+### Rinominare i metodi oscuri
+
+Ci sono metodi che hanno nomi degni di un codice morse ma che poco dicono sulle operazioni che vanno a intraprendere.
+Un buon inizio potrebbe essere quello di rinominarli in modo che chi legge abbia la possibilità di capire cosa fa quel metodo, senza necessità alcuna di andarlo a verificare.
+Prendiamo ad esempio questo metodo:
+
+```
+function getNetto() {
+  basePrice = cart.totalPrice();
+  return basePrice * discount();
+}
+```
+
+Leggendone il contenuto è chiaro che ci sta tornando il totale del carrello al netto di eventuali sconti ma, senza conoscerne il contenuto, come facciamo a esserne sicuri?
+Proviamo quindi a rinominarlo con qualcosa di più "parlante".
+
+```
+function getAmountWithAppliedDiscount() {
+  basePrice = cart.totalPrice();
+  return basePrice * discount();
+}
+```
+
+Cambiandone il nome, abbiamo reso più chiaro cosa farà e rendiamo più semplice la vita a chi quel metodo dovrà usarlo.
+
+### Cambiare la firma dei metodi
+
+Partendo dal presupposto che il codice è vivo, si ci può trovare nelle condizioni in cui al metodo inizialmente venivano passate diverse informazioni, che magari ora non sono più necessarie. E magari non lo sono perché nel frattempo, quella parte di codice era stata già snellita in qualche attività precedente.
+Possiamo quindi rimuovere le informazioni in eccesso, modificandone la firma.
+Inoltre, nei linguaggi tipizzati, potremmo verificare se c'è la possibilità di rendere detta firma più stringente, ammettendo e/o restituendo tipi meno generici.
```

Comment:
Riporto il commento di Angelo:
Qui sottolineerei che è importante farlo solo nelle major updates, e che sarebbe preferibile utilizzare in prima battuta delle keywords o delle annotazioni speciali come @deprecated per comunicare a chi utilizza quei metodi che quei parametri non saranno più utilizzati a posteriori.

Reply:
Mi torna quanto dice Angelo.
Aggiungerei un paragrafo sotto a questi esprimendo che quando si lavora su un prodotto utilizzato da altre persone, specialmente se esterne alla propria azienda o al proprio team, sarebbe meglio adottare un approccio di deprecation nel quale prima si indicano come deprecate le firme dei metodi o parte di esse, e solo ad una data stabilita e comunicata si proceda con la rimozione.